File: Tests/Test_GetFileIcon.vb
Option Explicit On
Option Strict On
Option Compare Binary
Option Infer Off

Imports System
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO
Imports System.Security.Cryptography

Namespace Tests
    Public Module Tests_GetFileIcon
        Function Test_GetFileIcon1(rootTestFolder As String) As Boolean
            Dim ico As Icon = WalkmanLib.GetFileIcon(Environment.GetEnvironmentVariable("WinDir"))

            Using savedImage As New DisposableFile(Path.Combine(rootTestFolder, "getFileIcon1.png"), False, False)
                ico.ToBitmap().Save(savedImage)

                Return TestBoolean("GetFileIcon1", File.Exists(savedImage), True)
            End Using
        End Function

        Function Test_GetFileIcon2(rootTestFolder As String) As Boolean
            Dim ico As Icon = WalkmanLib.GetFileIcon(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "explorer.exe"))

            Using savedImage As New DisposableFile(Path.Combine(rootTestFolder, "getFileIcon2.png"), False, False)
                ico.ToBitmap().Save(savedImage)

                Return TestBoolean("GetFileIcon2", File.Exists(savedImage), True)
            End Using
        End Function

        Function Test_GetFileIcon3() As Boolean
            Dim ex As Exception = New NoException
            Try
                Dim ico As Icon = WalkmanLib.GetFileIcon("nonexistantfile.txt")
            Catch ex2 As Exception
                ex = ex2
            End Try

            Return TestType("GetFileIcon3", ex.GetType(), GetType(FileNotFoundException))
        End Function

        Function Test_GetFileIcon4() As Boolean
            Dim ex As Exception = New NoException
            Try
                Dim ico As Icon = WalkmanLib.GetFileIcon("nonexistantfile.txt", False)
            Catch ex2 As Exception
                ex = ex2
            End Try

            Return TestType("GetFileIcon4", ex.GetType(), GetType(NoException))
        End Function

        Private Function ImageIsEqual(bmp1 As Bitmap, bmp2 As Bitmap) As Boolean
            If bmp1.PixelFormat <> bmp2.PixelFormat Then Return False
            If Not bmp1.RawFormat.Equals(bmp2.RawFormat) Then Return False
            If bmp1.Size <> bmp2.Size Then Return False

            For x As Integer = 0 To bmp1.Width - 1
                For y As Integer = 0 To bmp1.Height - 1
                    If bmp1.GetPixel(x, y) <> bmp2.GetPixel(x, y) Then
                        Return False
                    End If
                Next
            Next

            Return True
        End Function

        Function Test_GetFileIcon5() As Boolean
            Dim ico1 As Icon = WalkmanLib.GetFileIcon(".txt", False)
            Dim ico2 As Icon = WalkmanLib.GetFileIcon(".txt", False)

            Return TestBoolean("GetFileIcon5", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), True)
        End Function

        Function Test_GetFileIcon6() As Boolean
            Dim ico1 As Icon = WalkmanLib.GetFileIcon(".txt", False, linkOverlay:=False)
            Dim ico2 As Icon = WalkmanLib.GetFileIcon(".txt", False, linkOverlay:=True)

            Return TestBoolean("GetFileIcon6", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), False)
        End Function

        Function Test_GetFileIcon7() As Boolean
            Dim filePath As String = Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe")
            Dim ico1 As Icon = WalkmanLib.GetFileIcon(filePath)
            Dim ico2 As Icon = WalkmanLib.ExtractIconByIndex(filePath, 0, 16)

            Return TestBoolean("GetFileIcon7", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), True)
        End Function

        Function Test_GetFileIcon8() As Boolean
            Dim filePath As String = Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe")
            Dim ico1 As Icon = WalkmanLib.GetFileIcon(filePath, smallIcon:=False)
            Dim ico2 As Icon = WalkmanLib.ExtractIconByIndex(filePath, 0, 32)

            Return TestBoolean("GetFileIcon8", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), True)
        End Function

        Private Function Sha1Image(img As Image) As String
            Using ms As New MemoryStream()
                ' ImageFormat.MemoryBmp has no encoder, but it's still a BMP...
                img.Save(ms, If(img.RawFormat.Equals(ImageFormat.MemoryBmp), ImageFormat.Bmp, img.RawFormat))

                ms.Position = 0
                Dim hash As Byte() = New SHA1Managed().ComputeHash(ms)
                Return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant()
            End Using
        End Function

        Function Test_GetFileIcon9() As Boolean
            ' the icon for %WinDir%\hh.exe hasn't changed since XP, should be pretty safe
            Using ico As Icon = WalkmanLib.GetFileIcon(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe")),
                    bmp As Bitmap = ico.ToBitmap()

                Dim hash As String = Sha1Image(bmp)

                Return TestString("GetFileIcon9", hash, "f805d57e484d7f00298fe32b072b8c57928b3edb")
            End Using
        End Function
    End Module
End Namespace

File: Tests/Test_GetFileIcon.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Security.Cryptography;

namespace Tests {
    public static class Tests_GetFileIcon {
        public static bool Test_GetFileIcon1(string rootTestFolder) {
            Icon ico = WalkmanLib.GetFileIcon(Environment.GetEnvironmentVariable("WinDir"));

            using (var savedImage = new DisposableFile(Path.Combine(rootTestFolder, "getFileIcon1.png"), false, false)) {
                ico.ToBitmap().Save(savedImage);
                return GeneralFunctions.TestBoolean("GetFileIcon1", File.Exists(savedImage), true);
            }
        }

        public static bool Test_GetFileIcon2(string rootTestFolder) {
            Icon ico = WalkmanLib.GetFileIcon(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "explorer.exe"));

            using (var savedImage = new DisposableFile(Path.Combine(rootTestFolder, "getFileIcon2.png"), false, false)) {
                ico.ToBitmap().Save(savedImage);

                return GeneralFunctions.TestBoolean("GetFileIcon2", File.Exists(savedImage), true);
            }
        }

        public static bool Test_GetFileIcon3() {
            Exception ex = new NoException();
            try {
                Icon ico = WalkmanLib.GetFileIcon("nonexistantfile.txt");
            } catch (Exception ex2) {
                ex = ex2;
            }

            return GeneralFunctions.TestType("GetFileIcon3", ex.GetType(), typeof(FileNotFoundException));
        }

        public static bool Test_GetFileIcon4() {
            Exception ex = new NoException();
            try {
                Icon ico = WalkmanLib.GetFileIcon("nonexistantfile.txt", false);
            } catch (Exception ex2) {
                ex = ex2;
            }

            return GeneralFunctions.TestType("GetFileIcon4", ex.GetType(), typeof(NoException));
        }

        private static bool ImageIsEqual(Bitmap bmp1, Bitmap bmp2) {
            if (bmp1.PixelFormat != bmp2.PixelFormat)
                return false;
            if (!bmp1.RawFormat.Equals(bmp2.RawFormat))
                return false;
            if (bmp1.Size != bmp2.Size)
                return false;

            for (int x = 0; x < bmp1.Width; x++) {
                for (int y = 0; y < bmp1.Height; y++) {
                    if (bmp1.GetPixel(x, y) != bmp2.GetPixel(x, y)) {
                        return false;
                    }
                }
            }

            return true;
        }

        public static bool Test_GetFileIcon5() {
            Icon ico1 = WalkmanLib.GetFileIcon(".txt", false);
            Icon ico2 = WalkmanLib.GetFileIcon(".txt", false);

            return GeneralFunctions.TestBoolean("GetFileIcon5", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), true);
        }

        public static bool Test_GetFileIcon6() {
            Icon ico1 = WalkmanLib.GetFileIcon(".txt", false, linkOverlay: false);
            Icon ico2 = WalkmanLib.GetFileIcon(".txt", false, linkOverlay: true);

            return GeneralFunctions.TestBoolean("GetFileIcon6", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), false);
        }

        public static bool Test_GetFileIcon7() {
            string filePath = Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe");
            Icon ico1 = WalkmanLib.GetFileIcon(filePath);
            Icon ico2 = WalkmanLib.ExtractIconByIndex(filePath, 0, 16);

            return GeneralFunctions.TestBoolean("GetFileIcon7", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), true);
        }

        public static bool Test_GetFileIcon8() {
            string filePath = Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe");
            Icon ico1 = WalkmanLib.GetFileIcon(filePath, smallIcon: false);
            Icon ico2 = WalkmanLib.ExtractIconByIndex(filePath, 0, 32);

            return GeneralFunctions.TestBoolean("GetFileIcon8", ImageIsEqual(ico1.ToBitmap(), ico2.ToBitmap()), true);
        }

        private static string Sha1Image(Image img) {
            using (var ms = new MemoryStream()) {
                // ImageFormat.MemoryBmp has no encoder, but it's still a BMP...
                img.Save(ms, img.RawFormat.Equals(ImageFormat.MemoryBmp) ? ImageFormat.Bmp : img.RawFormat);

                ms.Position = 0;
                byte[] hash = new SHA1Managed().ComputeHash(ms);
                return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant();
            }
        }

        public static bool Test_GetFileIcon9() {
            // the icon for %WinDir%\hh.exe hasn't changed since XP, should be pretty safe
            using (Icon ico = WalkmanLib.GetFileIcon(Path.Combine(Environment.GetEnvironmentVariable("WinDir"), "hh.exe")))
            using (Bitmap bmp = ico.ToBitmap()) {
                string hash = Sha1Image(bmp);

                return GeneralFunctions.TestString("GetFileIcon9", hash, "f805d57e484d7f00298fe32b072b8c57928b3edb");
            }
        }
    }
}
